### PR TITLE
Aaijazi error code for tablet errors

### PR DESCRIPTION
**Pull request by aaijazi**
@sougou @alainjobart 

You might want to start off by reading the new descriptions of the error codes in vtrpc.proto. This is close to what @sougou and I discussed, although I made a couple of key changes.

The main things to keep in mind:

- The old ErrRetry should become QUERY_NOT_SERVED
- The old ErrFatal is close to the new INTERNAL_ERROR (but not exactly)
- The old ErrFail has become split into a few different errors, mostly: UNKNOWN_ERROR (errors from MySQL), BAD_INPUT, or PERMISSION_DENIED. 

I'm definitely willing to discuss what categories some of these errors should go into, there's still a couple of big question marks. I identified those with TODOs.

### CODE REVIEW

**Comment by alainjobart** [comment]
This is cool! Lots of detailed work, thanks for doing it :)

Probably Sugu should review it in details to make sure the codes are correct...

**Comment by sougou** [comment]
LGTM
I agree with converting most of the ErrFatal to INTERNAL_ERROR. But I'm slightly skeptical about the possibly transient ones, like MySQL going down unexpectedly.
It boils down to how we'll react to those errors. Would we retry the query on an INTERNAL_ERROR? If yes, then we're good. If we're not, then INTERNAL_ERROR won't work well for the case where MySQL went down unexpectedly, because that's a query we could have retried elsewhere.
I'm still ok with submitting this because we're not going to get this right on the first attempt. So, we just fix forward based on experience (past & future).

File: go/vt/tabletserver/dbconn.go [hunk 1]
```diff
@@ -58,17 +59,21 @@ func (dbc *DBConn) Exec(ctx context.Context, query string, maxrows int, wantfiel
...
 		}
 		err2 := dbc.reconnect()
 		if err2 != nil {
 			go checkMySQL()
-			return nil, NewTabletErrorSql(ErrFatal, err)
+			return nil, NewTabletErrorSql(ErrFatal, vtrpc.ErrorCode_INTERNAL_ERROR, err)
 		}
 	}
-	return nil, NewTabletErrorSql(ErrFatal, errors.New("dbconn.Exec: unreachable code"))
+	return nil, NewTabletErrorSql(ErrFatal, vtrpc.ErrorCode_INTERNAL_ERROR, errors.New("dbconn.Exec: unreachable code"))
```
Comment (sougou) [inline go/vt/tabletserver/dbconn.go]: The go convention is to use simple panics for unreachable code.

Comment (alainjobart) [inline go/vt/tabletserver/dbconn.go]: I thought the compiler was good enough to detect it now, and you don't even need anything?

Comment (sougou) [inline go/vt/tabletserver/dbconn.go]: If compiler allows it, that's even better.

Comment (enisoc) [inline go/vt/tabletserver/dbconn.go]: If the compiler doesn't allow it, you could also put `break` in `case attempt == 2:` and move the return below the loop. I could argue that's easier for humans to understand too. It's the code that gets executed after you've used up all your attempts.

Comment (enisoc) [inline go/vt/tabletserver/dbconn.go]: On second thought, you'd need a label on the for loop to do that, which some people find ugly.

**Comment by aaijazi** [comment]
@sougou: to answer your question, I'm considering INTERNAL_ERROR to be retryable by VtGate, but not something that clients should retry. INTERNAL_ERRORs, for the most part, only make sense to retry on a different VtTablet. In essence, I don't think clients should control which VtTablet exactly they want to route a request to, but VtGate can.

This somewhat falls apart, because of two cases:

1. Transient MySQL connection issues. I can see the case for these being considered TRANSIENT_ERROR instead of INTERNAL_ERROR, but I decided that I'd rather have clients delegate the retry behavior to VtGate, and not retry themselves.

2. There are some INTERNAL_ERRORs that are global across the VtTablet fleet, and thus shouldn't be retried by VtGate. Unfortunately, there's no better error code than INTERNAL_ERROR, so we're going to end up retrying a few more errors than we previously used to.

File: go/vt/tabletserver/dbconn.go (hunk 1 above)
Comment (aaijazi) [inline go/vt/tabletserver/dbconn.go]: I know the compiler catches some unreachable code paths, but it seems like it doesn't work for everything. It doesn't catch this case.

I'm going to leave this code as-is.

**Pull request merged**